### PR TITLE
Codemirror

### PR DESCRIPTION
**Pull request by holgerkoser**
**What this PR does / why we need it**:
This PR replaces the readonly shoot yaml tab and the fullscreen shoot yaml editor with a single shoot yaml editor page.
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement user
The shoot yaml editor page opens in readonly mode. It has a bottom toolbar with buttons to switch to edit mode or back to readonly mode. On the right there are buttons to download and copy the shoot resource. In edit mode there are additional buttons to save, reload the shoot resource and to undo and redo changes. The user can now add, change and remove annotations and labels as long as they are not in the reserved 'garden.sapcloud.io' namespace. Only the annotations 'shoot.garden.sapcloud.io/ignore' and  'garden.sapcloud.io/purpose' from this reserved namespace can be modified. 
```

### CODE REVIEW

**Comment by petersutter** [comment]
when editing the yaml and pressing save the cursor jumps to the start of the resource

**Comment by petersutter** [comment]
editor should be dirty after first change event until refresh or save, independent of editor state
change tooltip to `Discard Changes and Reload`

**Comment by petersutter** [comment]
never save if there are no changes

**Comment by petersutter** [comment]
remove trailing dot in single sentence tooltip

**Comment by petersutter** [comment]
remove replace spec in backend

File: frontend/src/router/index.js [hunk 1]
```diff
@@ -76,6 +77,26 @@ export default function createRouter ({store, userManager}) {
```
Comment (petersutter) [inline frontend/src/router/index.js]: add docu for tabs

File: frontend/src/layouts/Default.vue [hunk 2]
```diff
@@ -19,26 +19,54 @@ limitations under the License.
...
+      if (wrapElement && wrapElement.style) {
+        assign(wrapElement.style, {
+          overflowY: 'hidden'
+        })
+      }
     }
   }
 </script>
+
+<style lang="styl" scoped>
```
Comment (petersutter) [inline frontend/src/layouts/Default.vue]: remove style block

File: frontend/src/layouts/Default.vue [hunk 3]
```diff
@@ -19,26 +19,54 @@ limitations under the License.
...
+      MainContent,
       Loading,
       GSnotify
+    },
+    beforeRouteUpdate (to, from, next) {
+      this.$refs.content.setScrollTop(0)
+      next()
+    },
+    mounted () {
+      const element = this.$el
```
Comment (petersutter) [inline frontend/src/layouts/Default.vue]: move to method

File: frontend/src/pages/ShootItemEditor.vue [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+      this.destroyInstance()
+    }
+  }
+</script>
+
+<style lang="styl" scoped>
+  .no-margin
+    margin: 0 !important
+  .position-relative
+    position: relative !important
```
Comment (petersutter) [inline frontend/src/pages/ShootItemEditor.vue]: check if still needed

File: frontend/src/pages/ShootItemEditor.vue [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+            </div>
+            <span v-if="hasConflict">Cluster resource has been modified<br>by another user or process</span>
+            <span v-else>Cluster resource can be saved<br>without any conflicts</span>
+          </v-tooltip>
+        </v-flex>
+      </v-layout>
+    </v-flex>
+    <v-dialog v-model="dialog" persistent scrollable max-width="360px" @keydown.esc="resolveAction(false)">
+      <v-card>
+        <v-card-title primary-title class="orange grey--text text--lighten-4 headline" v-text="action.title"></v-card-title>
```
Comment (petersutter) [inline frontend/src/pages/ShootItemEditor.vue]: orange darken-2

File: frontend/src/pages/ShootItemEditor.vue [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+        <v-divider></v-divider>
+        <v-card-text style="height: 120px;" v-html="action.text"></v-card-text>
```
Comment (petersutter) [inline frontend/src/pages/ShootItemEditor.vue]: reduce height to 80px

File: frontend/src/pages/ShootItemEditor.vue [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+    </v-snackbar>
+  </v-layout>
+</template>
+
+<script>
+  import { mapGetters } from 'vuex'
+  import { replaceShoot } from '@/utils/api'
+  import Clipboard from 'clipboard'
+  import download from 'downloadjs'
+  import Popper from 'vue-popperjs'
```
Comment (petersutter) [inline frontend/src/pages/ShootItemEditor.vue]: remove

File: frontend/src/pages/ShootItemEditor.vue [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+          }
+        }
+        this.$instance = undefined
+      },
+      setOption (key, value) {
+        if (this.$instance) {
+          this.$instance.setOption(key, value)
+        }
+      },
+      getOption (key) {
```
Comment (petersutter) [inline frontend/src/pages/ShootItemEditor.vue]: remove (currently) unused setOption and getOption

File: frontend/src/pages/ShootItemEditor.vue [hunk 9]
```diff
@@ -0,0 +1,485 @@
...
+    },
+    mounted () {
+      const value = this.$localStorage.getItem('showShootEditorWarning')
+      this.modificationWarning = value === null || value === 'true'
+      this.createInstance(this.$refs.container)
+      this.update(this.value)
+      this.refresh()
+      // clipboard
+      const vm = this
+      const clipboard = new Clipboard(vm.$refs.btnCopy.$el, {
```
Comment (petersutter) [inline frontend/src/pages/ShootItemEditor.vue]: maybe use target

File: frontend/src/pages/ShootItemEditor.vue [hunk 10]
```diff
@@ -0,0 +1,485 @@
...
+          }
+        }
+      }
+    },
+    async beforeRouteLeave (to, from, next) {
+      if (this.clean) {
+        return next()
+      }
+      try {
+        next(await this.confirmNavigation())
```
Comment (petersutter) [inline frontend/src/pages/ShootItemEditor.vue]: two lines

File: frontend/src/utils/api.js [hunk 11]
```diff
@@ -83,6 +83,10 @@ export function deleteShoot ({namespace, name, user}) {
   return deleteResource(`/api/namespaces/${namespace}/shoots/${name}`, user)
 }
 
+export function replaceShoot ({namespace, name, user, data}) {
+  return updateResource(`/api/namespaces/${namespace}/shoots/${name}`, user, data)
+}
+
 export function replaceShootSpec ({namespace, name, user, data}) {
```
Comment (petersutter) [inline frontend/src/utils/api.js]: remove

File: frontend/src/router/index.js (hunk 1 above)
Comment (holgerkoser) [reply frontend/src/router/index.js]: :heavy_check_mark:

**Comment by holgerkoser** [comment]
> remove replace spec in backend

:heavy_check_mark:

File: frontend/src/layouts/Default.vue (hunk 2 above)
Comment (holgerkoser) [reply frontend/src/layouts/Default.vue]: :heavy_check_mark:

File: frontend/src/layouts/Default.vue (hunk 3 above)
Comment (holgerkoser) [reply frontend/src/layouts/Default.vue]: :heavy_check_mark:

File: frontend/src/pages/ShootItemEditor.vue (hunk 4 above)
Comment (holgerkoser) [reply frontend/src/pages/ShootItemEditor.vue]: It is needed for the absolute snackbar

File: frontend/src/pages/ShootItemEditor.vue (hunk 5 above)
Comment (holgerkoser) [reply frontend/src/pages/ShootItemEditor.vue]: :heavy_check_mark:

File: frontend/src/pages/ShootItemEditor.vue (hunk 6 above)
Comment (holgerkoser) [reply frontend/src/pages/ShootItemEditor.vue]: :heavy_check_mark:

File: frontend/src/pages/ShootItemEditor.vue (hunk 7 above)
Comment (holgerkoser) [reply frontend/src/pages/ShootItemEditor.vue]: :heavy_check_mark:

File: frontend/src/pages/ShootItemEditor.vue (hunk 8 above)
Comment (holgerkoser) [reply frontend/src/pages/ShootItemEditor.vue]: :heavy_check_mark:

File: frontend/src/pages/ShootItemEditor.vue (hunk 9 above)
Comment (holgerkoser) [reply frontend/src/pages/ShootItemEditor.vue]: target determines from which element the content to be copied is taken. We provide the content in the text callback. The button still has the focus after clicking. I now focus the editor after all toolbar buttons clicks.

File: frontend/src/pages/ShootItemEditor.vue (hunk 10 above)
Comment (holgerkoser) [reply frontend/src/pages/ShootItemEditor.vue]: :heavy_check_mark:

File: frontend/src/utils/api.js (hunk 11 above)
Comment (holgerkoser) [reply frontend/src/utils/api.js]: :heavy_check_mark:

File: frontend/src/pages/ShootItemEditor.vue (hunk 9 above)
Comment (holgerkoser) [reply frontend/src/pages/ShootItemEditor.vue]: :heavy_check_mark:

File: frontend/src/pages/ShootItemEditor.vue (hunk 4 above)
Comment (holgerkoser) [reply frontend/src/pages/ShootItemEditor.vue]: :heavy_check_mark:

**Pull request merged**